content/source/docs/extend/resources.html.md

```diff
@@ -23,6 +23,145 @@ A key component to Provider development is defining a resource. Aspects of this
+- Change the existing resource `Schema` as necessary.
+- If the `SchemaVersion` field for the resource is already defined, increase its value by one. If `SchemaVersion` is not defined for the resource, add `SchemaVersion: 1` to the resource (resources default to `SchemaVersion: 0` if undefined).
+- Implement the `StateUpgraders` field for the resource, which is a list of `StateUpgrade`. The new `StateUpgrade` should be configured with the following:
+  - `Type` set to `CoreConfigSchema().ImpliedType()` of the saved `schema.Resource` function above.
+  - `Upgrade` set to a function that modifies the attribute(s) appropriately for the migration.
+  - `Version` set to the version of the schema before this migration. If no previous state migrations were performed, this should be set to `0`.
+
+For example, with a resource without previous state migrations:
+
+```go
+package example
+
+import "github.com/hashicorp/terraform/helper/schema"
+
+func resourceExampleInstance() *schema.Resource {
+    return &schema.Resource{
+        Create: resourceExampleInstanceCreate,
+        Read:   resourceExampleInstanceRead,
+        Update: resourceExampleInstanceUpdate,
+        Delete: resourceExampleInstanceDelete,
+
+        Schema: map[string]*schema.Schema{
+            "name": {
+                Type:     schema.TypeString,
+                Required: true,
+            },
+        },
+    }
+}
+```
+
+Say the `instance` resource API now requires the `name` attribute to end with a period `"."`
+
+```go
+package example
+
+import (
+    "fmt"
+    "strings"
+
+    "github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceExampleInstance() *schema.Resource {
+    return &schema.Resource{
+        Create: resourceExampleInstanceCreate,
+        Read:   resourceExampleInstanceRead,
+        Update: resourceExampleInstanceUpdate,
+        Delete: resourceExampleInstanceDelete,
+
+        Schema: map[string]*schema.Schema{
+            "name": {
+                Type:     schema.TypeString,
+                Required: true,
+                ValidateFunc: func(v interface{}, k string) (warns []string, errs []error) {
+                    if !strings.HasSuffix(v.(string), ".") {
+                        errs = append(errs, fmt.Errorf("%q must end with a period '.'", k))
+                    }
+                    return
+                },
+            },
+        },
+        SchemaVersion: 1,
+        StateUpgraders: []schema.StateUpgrader{
+            {
+                Type:    resourceExampleInstanceResourceV0().CoreConfigSchema().ImpliedType(),
+                Upgrade: resourceExampleInstanceStateUpgradeV0,
+                Version: 0,
+            },
+        },
+    }
+}
+
+func resourceExampleInstanceResourceV0() *schema.Resource {
```

Comment:
I wish there was a good/easy way to reference the _current_ schema, instead of copying or moving it into a function/variable. Specifically in this scenario the only difference is the validation function, which isn't significant, so we could reference the same schema, right?

Not a big deal, but I can imagine some people following the guide too literally and copying schema when it doesn't need to be copied.

Reply:
The risk of referencing the "current" schema is that while it might be correct at the time the migration was written, it's likely to keep evolving over time and may become incorrect. Copying it is definitely a heavy way to work around that, but it seems safer than assuming that the current schema is correct and then having to remember to copy it later if there's a subsequent breaking change that _does_ require the prior schema snapshot.